sys/linux: add IMA/EVM interfaces

evdenis opened the pull request:
It's better to disable syscall write$evm. An attempt to write to "/sys/kernel/security/integrity/evm/evm" without additional vm setup highly likely will result in "lost connection to test machine" error due to -ENOKEY error. Recommended boot args for fuzzing: "ima_appraise=fix evm=fix"

evdenis committed:
sys/linux: describe integrity interfaces

Move integrity descriptions from filesystem.txt since they aren't
valid mount options.

Add descriptions for:
 - /sys/kernel/security/ima/*
 - /sys/kernel/security/evm/*

openat$evm is marked as disabled. An attempt to write to
"/sys/kernel/security/integrity/evm/evm" without additional vm setup
highly likely will result in "lost connection to test machine" error
due to -ENOKEY error.

Recommended boot args for fuzzing: "ima_appraise=fix evm=fix"

Possible boot args for integrity:
 - ima_appraise={ off | enforce | fix | log }
 - ima_hash={ md5 | sha1 | rmd160 | sha256 | sha384 | sha512 | ... }
 - ima_policy="tcb | appraise_tcb | secure_boot | fail_securely"
 - ima_template={ ima | ima-ng | ima-sig }
 - ima_template_fmt
 - ima.ahash_minsize, ima.ahash_bufsize
 - evm={ fix }

Signed-off-by: Denis Efremov <efremov@linux.com>

evdenis committed:
sys/linux: update IMA mount options

Add keyrings, appraise_flag, template. Update ima_funcs, appraise_type
values.

Signed-off-by: Denis Efremov <efremov@linux.com>

evdenis commented on sys/linux/security_ima.txt:
```diff
@@ -0,0 +1,112 @@
+] [varlen]
+
+resource fd_ima_policy[fd]
+openat$ima_policy(fd const[AT_FDCWD], file ptr[in, string["/sys/kernel/security/integrity/ima/policy"]], flags const[O_RDWR], mode const[0]) fd_ima_policy
```
For some reason I get:
```
2020/07/24 20:15:11 disabling openat$ima_policy: open(/sys/kernel/security/integrity/ima/policy) failed: permission denied
```

However, this file is readable and writeable when I manually boot the same vm:
```
root@syzkaller:~# ls -la /sys/kernel/security/integrity/ima/policy
-rw------- 1 root root 0 Jul 24 17:39 /sys/kernel/security/integrity/ima/policy
echo dont_measure > /sys/kernel/security/integrity/ima/policy
[  103.093986][ T5729] ima: policy update completed
[  103.094230][   T28] audit: type=1807 audit(1595612451.977:2): action=dont_measure re1
[  103.098757][   T28] audit: type=1802 audit(1595612451.977:3): pid=5729 uid=0 auid=0 1
root@syzkaller:~# cat /sys/kernel/security/integrity/ima/policy
dont_measure
```

dvyukov replied:
> For some reason I get:
> 
> ```
> 2020/07/24 20:15:11 disabling openat$ima_policy: open(/sys/kernel/security/integrity/ima/policy) failed: permission denied
> ```
> 
> However, this file is readable and writeable when I manually boot the same vm:
> 
> ```
> root@syzkaller:~# ls -la /sys/kernel/security/integrity/ima/policy
> -rw------- 1 root root 0 Jul 24 17:39 /sys/kernel/security/integrity/ima/policy
> echo dont_measure > /sys/kernel/security/integrity/ima/policy
> [  103.093986][ T5729] ima: policy update completed
> [  103.094230][   T28] audit: type=1807 audit(1595612451.977:2): action=dont_measure re1
> [  103.098757][   T28] audit: type=1802 audit(1595612451.977:3): pid=5729 uid=0 auid=0 1
> root@syzkaller:~# cat /sys/kernel/security/integrity/ima/policy
> dont_measure
> ```

Weird.  You may add some logging to isSupportedOpenAt in pkg/host/syscalls_linux.go and then run with -debug flag.

dvyukov commented on sys/linux/filesystem.txt:
```diff
@@ -192,8 +189,24 @@ uuid_str {
+} [packed]
+
+ima_keyring_list {
+	list	optional[array[ima_opt_keyring[stringnoz[ima_keyrings]]]]
```
Nit: array effectively implies optional as it can have 0 elements.

dvyukov commented on sys/linux/filesystem.txt:
```diff
@@ -192,8 +189,24 @@ uuid_str {
+ima_appraise_type = "imasig", "imasig|modsig"
+ima_template = "ima", "ima-ng", "ima-sig", "ima-buf", "ima-modsig"
+# from security/integrity/digsig.c keyring_name
+# TODO: other values? .persistent_register, .dns_resolver?
```
@ebiggers do you know anything about this by any chance?

ebiggers replied:
Not very much.  If this API is accepting a keyring name (a.k.a. a "key description"), it might be good to add `key_desc` from `sys/linux/key.txt`.

evdenis replied:
key_desc contains '\0' in the end. I can't use it unmodified in policy because of this.
```
key_desc {
prefix  stringnoz["syz"]
id      proc[' ', 4, int8]
z       const[0, int8]
}
```
I can split the structure to key_desc and key_descz to be able to use it in security_ima.txt

dvyukov replied:
Splitting key_desc looks reasonable.

dvyukov commented:
> An attempt to write to "/sys/kernel/security/integrity/evm/evm" without additional vm setup highly likely will result in "lost connection to test machine" error due to -ENOKEY error.

Will ENOKEY be returned for all subsequent filesystem operations? Or why will it result in "lost connection"?

What is the additional VM setup? Is it these "ima_appraise=fix evm=fix"?
If it provokes "lost connection" by default all the time, it's not very good for syzbot and all other users out there. We try to make the default behavior without "special tuning by an expert required" sane. It won't scale for 1000 of kernel subsystems, nobody is an expert in 1000 kernel subsystems.

dvyukov commented:
> Recommended boot args for fuzzing: "ima_appraise=fix evm=fix"

If it's recommended for all setups, please add it to:
tools/create-gce-image.sh
tools/create-image.sh

If it's "weakly recommended", but it not really necessary, then add it to (for syzbot):
dashboard/config/upstream-*.cmdline

dvyukov commented on sys/linux/security_ima.txt:
```diff
@@ -0,0 +1,111 @@
+
+evm_xattr_list {
+	list	array[evm_xattr_line]
+	z	const[0, int8]
```
Not that it matters much, but I wonder if 0 is really required. Generally kernel code should look at the size of the write operation and 0-terminate itself if it needs the 0.

dvyukov commented:
The descriptions themselves look good to me. Though I don't understand any of these interface.

evdenis commented:
> Will ENOKEY be returned for all subsequent filesystem operations? Or why will it result in "lost connection"?

Yes,  for all subsequent "write" filesystem operations.

> What is the additional VM setup? Is it these "ima_appraise=fix evm=fix"?

No, one needs to generate and load crypto keys. Will it be better to just comment line?
```
# write$evm(fd fd_evm, buf ptr[in, fmt[dec, flags[evm_flags]]], count len[buf])
```

> If it's "weakly recommended", but it not really necessary, then add it to (for syzbot):
> dashboard/config/upstream-*.cmdline

By default IMA/EVM is starting in off mode. Switching them to fix modes should cover mode kernel operations like checksum computations for files/metadata, xattr updates etc.

evdenis committed:
dashboard/config: set ima_appraise=fix, evm=fix

Set ima_appraise=fix, evm=fix for integrity module to
generate checksums and update xattrs.

Signed-off-by: Denis Efremov <efremov@linux.com>

dvyukov commented:
> No, one needs to generate and load crypto keys.

How hard is it to do? What what does it involve?

> Will it be better to just comment line?

I think we need to disable the open call, otherwise syzkaller will find gazillion other ways of how to write to it.
It can now be done by adding `(disabled)` attribute to open and write, this way they are still parsed and checked for correctness.

evdenis commented:
> How hard is it to do? What what does it involve?

1. Generate keys in create-image.sh
```
$ dd if=/dev/urandom bs=1 count=32 status=none | keyctl padd user kmk-user @u
$ keyctl pipe `keyctl search @u user kmk-user` > /etc/keys/kmk-user.blob
$ keyctl add encrypted evm-key "new user:kmk-user 32" @u
$ keyctl pipe `keyctl search @u encrypted evm-key` > /etc/keys/evm-user.blob
```

2. Update init scripts to load keys during boot with commands like:
```
keyctl show | grep -q kmk-user || keyctl add user kmk-user "`cat /etc/keys/kmk-user.blob`" @u > /dev/null
keyctl add encrypted evm-key "load `cat /etc/keys/evm-user.blob`" @u > /dev/null
ima_id=`keyctl newring _ima @u` && /usr/bin/evmctl import --rsa /etc/keys/rsa_public.pem $ima_id > /dev/null
evm_id=`keyctl newring _evm @u` && /usr/bin/evmctl import --rsa /etc/keys/rsa_public.pem $evm_id > /dev/null
```
3. At this step it's possible to write to evm file, e.g. "echo 1 > /sys/kernel/security/integrity/evm/evm"

dvyukov commented:
> > How hard is it to do? What what does it involve?
> 
> 1. Generate keys in create-image.sh
> 
> ```
> $ dd if=/dev/urandom bs=1 count=32 status=none | keyctl padd user kmk-user @u
> $ keyctl pipe `keyctl search @u user kmk-user` > /etc/keys/kmk-user.blob
> $ keyctl add encrypted evm-key "new user:kmk-user 32" @u
> $ keyctl pipe `keyctl search @u encrypted evm-key` > /etc/keys/evm-user.blob
> ```
> 
> 1. Update init scripts to load keys during boot with commands like:
> 
> ```
> keyctl show | grep -q kmk-user || keyctl add user kmk-user "`cat /etc/keys/kmk-user.blob`" @u > /dev/null
> keyctl add encrypted evm-key "load `cat /etc/keys/evm-user.blob`" @u > /dev/null
> ima_id=`keyctl newring _ima @u` && /usr/bin/evmctl import --rsa /etc/keys/rsa_public.pem $ima_id > /dev/null
> evm_id=`keyctl newring _evm @u` && /usr/bin/evmctl import --rsa /etc/keys/rsa_public.pem $evm_id > /dev/null
> ```
> 
> 1. At this step it's possible to write to evm file, e.g. "echo 1 > /sys/kernel/security/integrity/evm/evm"

Thanks!
Let's wait for @ebiggers, I think it's close to his area of expertise.

ebiggers commented:
> Let's wait for @ebiggers, I think it's close to his area of expertise.

Not that close.  I don't work on IMA or EVM.

dvyukov commented:
> > Let's wait for @ebiggers, I think it's close to his area of expertise.
> 
> Not that close. I don't work on IMA or EVM.

Ack. Thanks for the review.

I started reading:
https://sourceforge.net/p/linux-ima/wiki/Home/
which seems to be the description of the setup needed. But I feel it will take me days to grasp it all, figure out the details and test changes needed to deploy this on syzbot. I filed #1977 for now.

Denis, is it possible to understand if writing to the file will cause all writes to fail or not (is the key has been setup or not)? If yes, then I think the best solution is to detect presence of the key here:
https://github.com/google/syzkaller/blob/cb93dc6ac64225e09f44bac6c6cce1dae1b248b0/pkg/host/syscalls_linux.go#L24
where we check if we should enable/disable particular syscalls. Then we can leave it enabled and it will automatically work on any machines where they key is setup.

evdenis commented:
How to reproduce ENOKEY error:
```
root@syzkaller:~# echo 2 > /sys/kernel/security/integrity/evm/evm
root@syzkaller:~# sleep 20 && touch test.txt
[  526.976855][ T5771] evm: HMAC key is not set
[  526.977892][ T5771] evm: init_desc failed
touch: cannot touch 'test.txt': Required key not available
```

evdenis commented:
> Denis, is it possible to understand if writing to the file will cause all writes to fail or not (is the key has been setup or not)? If yes, then I think the best solution is to detect presence of the key here:
> https://github.com/google/syzkaller/blob/cb93dc6ac64225e09f44bac6c6cce1dae1b248b0/pkg/host/syscalls_linux.go#L24
> where we check if we should enable/disable particular syscalls. Then we can leave it enabled and it will automatically work on any machines where they key is setup.

I think we can check /proc/keys:
```
$ cat /proc/keys | grep evm
007c6d0c I--Q---     1 perm 3f010000     0     0 encrypted evm-key: 139
206619df I------     1 perm 1f0f0000     0     0 keyring   .evm: empty
```

String "encrypted evm-key" to detect the presence of hmac key which is requested in evm_init_key (https://github.com/torvalds/linux/blob/92ed301919932f777713b9172e525674157e983d/security/integrity/evm/evm_crypto.c#L357)

evdenis commented on sys/linux/filesystem.txt:
```diff
@@ -146,14 +146,7 @@ fs_options_security [
 	smackfstransmute	fs_opt_str["smackfstransmute"]
 
 # IMA options:
```
@dvyukov, do you want me to remove IMA mount options and move remaining structures to security_ima.txt?

ebiggers replied:
I think that's the only thing that makes sense.  `fs_options_security` is for mount options, which syzkaller passes to `mount()`.  But these aren't mount options but rather part of the syntax for specifying an IMA policy, to be written to `/sys/kernel/security/integrity/ima/policy`.

evdenis commented:
> How to reproduce ENOKEY error:
> 
> ```
> root@syzkaller:~# echo 2 > /sys/kernel/security/integrity/evm/evm
> root@syzkaller:~# touch test.txt # should be new file
> [  526.976855][ T5771] evm: HMAC key is not set
> [  526.977892][ T5771] evm: init_desc failed
> touch: cannot touch 'test.txt': Required key not available
> ```

2 is for "Enable digital signature validation" according to Documentation/ABI/testing/evm and I think that kernel should not accept the value without checking that hmac and digsig keys are loaded (pretty much as with 1).

evdenis commented:
To check that at least one digsig key loaded to _evm keyring during boot we can check:
```
$ evm_id=`keyctl newring _evm @u`
$ evmctl import /etc/keys/pubkey_ima.pem $evm_id
$ cat /proc/keys | grep keyring | grep _evm
1e8e1965 I--Q---     1 perm 3f010000     0     0 keyring   _evm: 1

# no keys
$ keyctl newring _test @u
$ cat /proc/keys | grep keyring | grep _test
0a9b9f6a I--Q---     1 perm 3f010000     0     0 keyring   _test: empty
```

evmctl uses add_key syscall:
```
# keyring creation
add_key("keyring", "_test1", NULL, 0, KEY_SPEC_USER_KEYRING) = 458492488

# key import
openat(AT_FDCWD, "/etc/keys/pubkey_evm.pem", O_RDONLY) = 3
futex(0x7fdbfafee6c4, FUTEX_WAKE_PRIVATE, 2147483647) = 0
fstat(3, {st_mode=S_IFREG|0644, st_size=272, ...}) = 0
read(3, "-----BEGIN PUBLIC KEY-----\nMIGfM"..., 4096) = 272
futex(0x7fdbfafee65c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
close(3)                                = 0
add_key("user", "\4ZR0\376\177", "\1\0\0\0\0\0\2\4\0\276b\216\270|\233\366\4\247\275e\230rq\324,\32\347;T\220V\251"..., 142, 177971050) = 327167649
```

evdenis committed:
sys/linux: split key_desc struct to key_desc and key_descz

It's handly to use key_desc struct in other rules, e.g. IMA/EVM.
In order to be able to do it we need to separate '\0' terminator
to a new structure.

Signed-off-by: Denis Efremov <efremov@linux.com>

evdenis committed:
sys/linux: use key_desc keyrings in security_ima.txt

Use syzbot keyrings in IMA/EVM policy.

Signed-off-by: Denis Efremov <efremov@linux.com>
